# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/xhr/XMLHttpRequestWorker.h
# Commit: 51bd2ff91c1a
# Full Hash: 51bd2ff91c1a4fd7f12dc5b4062be8657e82f8ee
# Author: Thomas Wisniewski <twisniewski@mozilla.com>
# Date: 2024-06-21 10:09:55
# Regressor Bug: 1842970
# File Overlap Count: 1
# Description:
#   Bug 1842970 - fire abort events more in line with the spec for worker XMLHttpRequest; r=valentin,sunil,dom-worker-reviewers,smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D201744
# ==============================================================================

diff -r f8d683b59147 -r 51bd2ff91c1a dom/xhr/XMLHttpRequestWorker.h
--- a/dom/xhr/XMLHttpRequestWorker.h	Fri Jun 21 06:11:46 2024 +0300
+++ b/dom/xhr/XMLHttpRequestWorker.h	Fri Jun 21 04:16:11 2024 +0000
@@ -48,14 +48,10 @@
 
   struct StateData {
     nsString mResponseURL;
-    uint32_t mStatus;
+    uint32_t mStatus{0};
     nsCString mStatusText;
-    uint16_t mReadyState;
-    bool mFlagSend;
-    nsresult mStatusResult;
-
-    StateData()
-        : mStatus(0), mReadyState(0), mFlagSend(false), mStatusResult(NS_OK) {}
+    uint16_t mReadyState{0};
+    nsresult mStatusResult{NS_OK};
   };
 
  private:
@@ -74,11 +70,13 @@
   JS::Heap<JSObject*> mResponseArrayBufferValue;
   JS::Heap<JS::Value> mResponseJSONValue;
 
+  uint32_t mEventStreamId{0};
   uint32_t mTimeout;
 
   bool mBackgroundRequest;
   bool mWithCredentials;
   bool mCanceled;
+  bool mFlagSend{false};  // spec flag
   bool mFlagSendActive;
 
   bool mMozAnon;
@@ -97,9 +95,7 @@
 
   void Unpin();
 
-  virtual uint16_t ReadyState() const override {
-    return mStateData->mReadyState;
-  }
+  virtual uint16_t ReadyState() const override;
 
   virtual void Open(const nsACString& aMethod, const nsAString& aUrl,
                     ErrorResult& aRv) override {
@@ -107,6 +103,8 @@
          aRv);
   }
 
+  uint32_t EventStreamId() const { return mEventStreamId; }
+
   virtual void Open(const nsACString& aMethod, const nsAString& aUrl,
                     bool aAsync, const nsAString& aUsername,
                     const nsAString& aPassword, ErrorResult& aRv) override {
@@ -239,9 +237,13 @@
 
   void MaybePin(ErrorResult& aRv);
 
-  void MaybeDispatchPrematureAbortEvents(ErrorResult& aRv);
+  void SetResponseToNetworkError();
 
-  void FireEvent(EventTarget* aTarget, const EventType& aEventType,
+  void RequestErrorSteps(ErrorResult& aRv,
+                         const ErrorProgressEventType& aEventType,
+                         nsresult aException = NS_ERROR_DOM_INVALID_STATE_ERR);
+
+  bool FireEvent(EventTarget* aTarget, const EventType& aEventType,
                  bool aUploadTarget, ErrorResult& aRv);
 
   void Send(JSContext* aCx, JS::Handle<JSObject*> aBody, ErrorResult& aRv);